internal/postgres/report.go

```diff
@@ -0,0 +1,496 @@
+	s := fmt.Sprintf(" (%s%% of %d rows written to Spanner)", pct(rows, badRows), rows)
+	switch {
+	case rows == 0:
+		return "NONE (no data rows found)"
+	case badRows == 0:
+		return fmt.Sprintf("EXCELLENT (all %d rows written to Spanner)", rows)
+	case good(rows, badRows):
+		return "GOOD" + s
+	case ok(rows, badRows):
+		return "OK" + s
+	default:
+		return "POOR" + s
+	}
+}
+
+func good(total, badCount int64) bool {
+	return badCount < total/20
+}
+
+func ok(total, badCount int64) bool {
+	return badCount < total/3
+}
+
+func rateConversion(rows, badRows, cols, warnings int64, missingPKey, summary bool) string {
+	return fmt.Sprintf("Schema conversion: %s.\n", rateSchema(cols, warnings, missingPKey, summary)) +
+		fmt.Sprintf("Data conversion: %s.\n", rateData(rows, badRows))
+}
+
+func generateSummary(conv *Conv, r []tableReport, badWrites map[string]int64) string {
+	cols := int64(0)
+	warnings := int64(0)
+	missingPKey := false
+	for _, t := range r {
+		weight := t.rows // Weight col data by how many rows in table.
+		if weight == 0 { // Tables without data count as if they had one row.
+			weight = 1
+		}
+		cols += t.cols * weight
+		warnings += t.warnings * weight
+		if t.syntheticPKey != "" {
+			missingPKey = true
+		}
+	}
+	// Don't use tableReport for rows/badRows stats because tableReport
+	// provides per-table stats for each table in the schema i.e. it omits
+	// rows for tables not in the schema. To handle this corner-case, use
+	// the source of truth for row stats: conv.stats.
+	rows := conv.Rows()
+	badRows := conv.BadRows() // Bad rows encountered during data conversion.
+	// Add in bad rows while writing to Spanner.
+	for _, n := range badWrites {
+		badRows += n
+	}
+	return rateConversion(rows, badRows, cols, warnings, missingPKey, true)
+}
+
+func ignoredStatements(conv *Conv) (l []string) {
+	for s := range conv.stats.statement {
+		switch s {
+		case "CreateFunctionStmt":
+			l = append(l, "functions")
+		case "CreateSeqStmt":
+			l = append(l, "sequences")
+		case "CreatePLangStmt":
+			l = append(l, "procedures")
+		case "CreateTrigStmt":
+			l = append(l, "triggers")
+		case "IndexStmt":
+			l = append(l, "(non-primary) indexes")
+		case "ViewStmt":
+			l = append(l, "views")
+		}
+	}
+	sort.Strings(l)
+	return l
+}
+
+func writeStmtStats(conv *Conv, w *bufio.Writer) {
+	type stat struct {
+		statement string
+		count     int64
+	}
+	var l []stat
+	for s, x := range conv.stats.statement {
+		l = append(l, stat{s, x.schema + x.data + x.skip + x.error})
+	}
+	// Sort by alphabetical order of statements.
+	sort.Slice(l, func(i, j int) bool {
+		return l[i].statement < l[j].statement
+	})
+	writeHeading(w, "Statements Processed")
+	w.WriteString("Analysis of statements in pg_dump output, broken down by statement type.\n")
+	w.WriteString("  schema: statements successfully processed for Spanner schema information.\n")
+	w.WriteString("    data: statements successfully processed for data.\n")
```

Comment:
Are these lines meant to be more indented than the others?

Reply:
Yes. I wanted these lines to line up so the key to the table would be easier to read.